src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage.cs

```diff
@@ -174,7 +174,18 @@ private void ReleaseConnection(SqlConnection connection)
                     return;
                 }
 
-                _connectionsPool.Push(connection);
+                try
+                {
+                    _connectionsPool.Push(connection);
+                }
+                catch
+                {
+                    // An exception (likely OutOfMemoryException) occurred while returning the connection to the pool.
+                    // The connection will be discarded, so make sure to close it so the finalizer doesn't crash the
+                    // process later.
+                    connection.Close_OnlyForUseBySqlPersistentStorage();
```

Comment:
It looks like this method allocates; is that worrisome?

Reply:
Are you sure?  why do you think that?

Reply:
No, because the allocation happens _after_ the call to `GC.SuppressFinalize`.

Reply:
@CyrusNajmabadi I see a foreach being used, so it'll allocate the enumerator (likely).

My concern here was we'll still be leaking the connection if that allocation were to throw. It feels like we really should be using a SafeHandle here or something to ensure that cleanup happens regardless?

Reply:
> @CyrusNajmabadi I see a foreach being used, so it'll allocate the enumerator (likely).

The GC.suppressfinalize happens first.  So we don't run the finalizer.

> My concern here was we'll still be leaking the connection if that allocation were to throw. It feels like we really should be using a SafeHandle here or something to ensure that cleanup happens regardless?

Sure.  That seems reasonable to me.  I don't have experience with these.  Do you have an example i can look at for using that?